token.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2024 The Godror Authors
+//
+//
+// SPDX-License-Identifier: UPL-1.0 OR Apache-2.0
+
+package godror
+
+/*
+#include <stdlib.h>
+#include <stdio.h>
+#include "dpiImpl.h"
+
+int TokenCallbackHandlerDebug(void* context, dpiAccessToken *token);
+
+*/
+import "C"
+
+import (
+	"context"
+	"github.com/godror/godror/dsn"
+	"log"
+	"sync"
+	"unsafe"
+)
+
+// AccessToken Callback information.
+type accessTokenCB struct {
+	ctx         context.Context
+	callback    func(context.Context, *dsn.AccessToken) error
+	id          uint64
+	ctoken      *C.char
+	cprivateKey *C.char
+}
+
+// Cannot pass *accessTokenCB to C, so pass an uint64 that points to
+// this map entry
+var (
+	accessTokenCBsMu sync.Mutex
+	accessTokenCBs   = make(map[uint64]*accessTokenCB)
```

Comment:
Use of runtime/cgo.Handle may be easier.

Reply:
I tried with cgo.Handle to store the callback context and hence avoid global map, accessTokenCBs . It is causing panic during `C.dpiPool_create`
`panic: runtime error: cgo argument has Go pointer to unpinned Go pointer`

I am able to simulate this panic with simple program.  Below program will panic but when we comment this line 
	`var h cgo.Handle`  so that we use global handle `h` , it works fine. 


```
package main

/*
struct dt {
  void *context;
};
typedef struct dt dt;

extern void MyGoPrint(void *context);
static inline void myprint(struct dt *a1) {
    MyGoPrint(a1->context);
}
*/
import "C"
import (
	"context"
	"runtime/cgo"
	"unsafe"
)

//export MyGoPrint
func MyGoPrint(context unsafe.Pointer) {
	h := *(*cgo.Handle)(context)
	val := h.Value().(accessTokenCB)
	println(val.id)
	h.Delete()
}

type At struct {
	Tok string
}

// AccessToken Callback information.
type accessTokenCB struct {
	ctx         context.Context
	callback    func(context.Context, *At) error
	id          uint64
	ctoken      *C.char
	cprivateKey *C.char
}

func getH() cgo.Handle {
	cb := func(ctx context.Context, tok *At) error {
		tok.Tok = "123"
		return nil
	}
	val := accessTokenCB{callback: cb, ctx: context.Background(), id: 32, ctoken: nil, cprivateKey: nil}
	h := cgo.NewHandle(val)
	return h

}

var h cgo.Handle

func main() {
	var h cgo.Handle
	h = getH()
	var poolCt C.dt
	poolCt.context = unsafe.Pointer(&h)
	C.myprint(&poolCt)
	// Output: 32
}

```

Reply:
```
package main

/*
extern void MyGoPrint(void *p);
static inline void myprint(void *p) {
    MyGoPrint(p);
}
*/
import "C"
import (
	"context"
	"runtime/cgo"
	"unsafe"
)

//export MyGoPrint
func MyGoPrint(p unsafe.Pointer) {
	h := *(*cgo.Handle)(p)
	val := h.Value().(accessTokenCB)
	println(val.id)
	h.Delete()
}

type At struct {
	Tok string
}

// AccessToken Callback information.
type accessTokenCB struct {
	ctx         context.Context
	callback    func(context.Context, *At) error
	id          uint64
	ctoken      *C.char
	cprivateKey *C.char
}

func getH() cgo.Handle {
	cb := func(ctx context.Context, tok *At) error {
		tok.Tok = "123"
		return nil
	}
	val := accessTokenCB{
		callback: cb, ctx: context.Background(),
		id: 32, ctoken: nil, cprivateKey: nil}
	h := cgo.NewHandle(val)
	return h

}

func main() {
	var h cgo.Handle
	h = getH()
	C.myprint(unsafe.Pointer(&h))
	// Output: 32
}
```

works (no fancy structs - that'd need C helper).

Reply:
Finding https://groups.google.com/g/golang-nuts/c/PLSIN5jmzpM

The rewritten code is even simpler (treat cgo.Handle as a number (C.uintptr_t)):

```
package main

/*
#include <stdint.h>

extern void MyGoPrint(uintptr_t p);
static inline void myprint(uintptr_t p) {
    MyGoPrint(p);
}
*/
import "C"
import (
	"context"
	"runtime/cgo"
)

//export MyGoPrint
func MyGoPrint(handle C.uintptr_t) {
	h := cgo.Handle(handle)
	val := h.Value().(accessTokenCB)
	println(val.id)
	h.Delete()
}

type At struct {
	Tok string
}

// AccessToken Callback information.
type accessTokenCB struct {
	ctx         context.Context
	callback    func(context.Context, *At) error
	id          uint64
	ctoken      *C.char
	cprivateKey *C.char
}

func getH() cgo.Handle {
	cb := func(ctx context.Context, tok *At) error {
		tok.Tok = "123"
		return nil
	}
	val := accessTokenCB{
		callback: cb, ctx: context.Background(),
		id: 32, ctoken: nil, cprivateKey: nil}
	h := cgo.NewHandle(val)
	return h

}

func main() {
	C.myprint(C.uintptr_t(getH()))
	// Output: 32
}
```

Reply:
In our case , we need to pass `poolCreateParams` struct which has  void * argument, `accessTokenCallbackContext`. Hence i cant use cgo.Handle right?

Reply:
Thanks . I made the changes by wrapping it now..